File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<vector<int>> v;
vector<int> visited;

void bfs(int start) {
    queue<int> q;
    q.push(start);
    visited[start] = true;
    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        for (int x: v[cur]) {
            if (!visited[x]) {
                visited[x] = true;
                q.push(x);
            }
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    v.assign(n, vector<int>());
    visited.assign(n, 0);

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (i != j && computers[i][j]) {
                v[i].push_back(j);
                v[j].push_back(i);
            }
        }
    }

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            answer++;
            bfs(i);
        }
    }

    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.kt
import java.util.LinkedList
import java.util.Queue 

class Solution {

    private lateinit var visited: BooleanArray
    private lateinit var info: Array<MutableList<Int>>

    fun solution(n: Int, computers: Array<IntArray>): Int {
        var answer = 0
        info = Array(n) { mutableListOf() }
        visited = BooleanArray(n) { false }

        for (i in 0 until n) {
            for (j in 0 until n) {
                if (i != j && computers[i][j] == 1) info[i].add(j)
            }
        }

        for (i in 0 until n) {
            if (!visited[i]) {
                bfs(i)
                answer++
            }
        }

        return answer
    }

    private fun bfs(start: Int) {
        val q: Queue<Int> = LinkedList()
        q.add(start)
        visited[start] = true
        while (q.isNotEmpty()) {
            val number = q.poll()
            for (i in info[number]) {
                if (!visited[i]) {
                    q.add(i)
                    visited[i] = true
                }
            }

        }
    }
}